apps/internal/options/options.go

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package options
+
+import (
+	"errors"
+	"fmt"
+)
+
+// CallOption implements an optional argument to a method call. See
+// https://blog.devgenius.io/go-call-option-that-can-be-used-with-multiple-methods-6c81734f3dbe
+// for an explanation of the usage pattern.
+type CallOption interface {
+	Do(any) error
+	callOption()
+}
+
+// ApplyOptions applies all the callOptions to options. options must be a pointer to a struct and
+// callOptions must be a list of objects that implement CallOption.
+func ApplyOptions[O, C any](options O, callOptions []C) error {
```

Comment:
Given the comment, you could use a constraint on `C`.  Or `C` isn't a generic parameter and is just `[]CallOption`?

Reply:
`C` is unconstrained to allow options of all types. For example, `AcquireTokenSilent` takes `...AcquireTokenSilentOption`. In practice those options will be anonymous structs implementing `CallOption`. However, `AcquireTokenSilentOption` itself doesn't implement `CallOption`. If `C` were simply a `[]CallOption`, passing `[]AcquireTokenSilentOption` would get a spurious compile error. You could call that error useful and the `any` a kludge. I can live with this in an internal API but am open to better ideas. The best alternative that comes to my mind would be to implement `CallOption` on the various option types, which in this design would mean adding a new public method to each of them or moving `CallOption` into `public` and `confidential` to avoid exporting that method.

Reply:
Does this mean there are cases where the type-assertion on line 23 fails and we expect this?

Reply:
No, it means the assertion never fails but I don't see a good way to obviate it.

Reply:
I get how this works now, and I agree that compromising on the internal API is a good choice (compared to the other options you listed).